File: leetcode_0098/java/leetcode_0098.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
public class leetcode_0098 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    class Solution {
        public boolean isValidBST(TreeNode root) {
            //检测合法BST，可用中序遍历来检测
            pre = null;
            return inOrder(root);
        }
        private TreeNode pre;
        private boolean inOrder(TreeNode root){
            if(root!=null){
                boolean leftValid = inOrder(root.left);
                if(!leftValid)
                    return false;
                if(pre == null)
                    pre = root;
                else if(pre.val >= root.val)
                    return false;
                else{
                    //pre.val < root.val
                    pre = root;
                }
                boolean rightValid = inOrder(root.right);
                if(!rightValid)
                    return false;
            }
            return true;
        }

        public List<List<Integer>> levelOrder(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<List<Integer>> res = new LinkedList();
            List<Integer> curEles = new LinkedList();
            while (cur.size() > 0) {
                for (int i = 0; i < cur.size(); i++) {
                    curEles.add(cur.get(i).val);
                    if (cur.get(i).left != null)
                        next.add(cur.get(i).left);
                    if (cur.get(i).right != null)
                        next.add(cur.get(i).right);
                }
                res.add(new LinkedList(curEles));
                cur = new LinkedList(next);
                curEles.clear();
                next.clear();
            }
            return res;
        }

        public List<String> levelOrderStr(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<String> res = new LinkedList();
            while (cur.size() > 0) {
                int flag = 0;
                for (int i = 0; i < cur.size(); i++) {
                    if (cur.get(i) == null) {
                        res.add("null");
                        continue;
                    } else
                        res.add(String.valueOf(cur.get(i).val));
                    TreeNode left = cur.get(i).left;
                    TreeNode right = cur.get(i).right;
                    if (left != null || right != null)
                        flag++;
                    next.add(left);
                    next.add(right);
                }
                cur = new LinkedList(next);
                if (flag == 0)
                    cur.clear();
                next.clear();
            }
            return res;
        }

        public <T> void printListArr(List<List<T>> list) {
            for (int i = 0; i < list.size(); i++) {
                if (i == list.size() - 1) {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.println(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " ] ");
                } else {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.printf(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " , ");
                }
            }
        }
    }
    public static void main(String[] args) {
        leetcode_0098.Solution solution = new leetcode_0098().new Solution();
        leetcode_0098.TreeNode root = new leetcode_0098().new TreeNode(5);
        leetcode_0098.TreeNode left = new leetcode_0098().new TreeNode(1);
        leetcode_0098.TreeNode right = new leetcode_0098().new TreeNode(4);
        leetcode_0098.TreeNode ll = null;
        leetcode_0098.TreeNode lr = null;
        leetcode_0098.TreeNode rl = new leetcode_0098().new TreeNode(3);
        leetcode_0098.TreeNode rr = new leetcode_0098().new TreeNode(6);
        left.left = ll;
        left.right = lr;
        right.left = rl;
        right.right = rr;
        root.left = left;
        root.right = right;
//      root = [5, 1, 4, null, null, 3, 6]
        List<String> rootStr = solution.levelOrderStr(root);
        System.out.println(Arrays.toString(rootStr.toArray(String[]::new)));
        /**
         * 解释: 输入为: [5,1,4,null,null,3,6]。
         *      根节点的值为 5 ，但是其右子节点值为 4 。
         */
        boolean validBST = solution.isValidBST(root);
        System.out.println(validBST);
        //result
        //false
    }
}

File: leetcode_0098/cpp/leetcode_0098.cpp
//
// Created by wangzy on 2022/4/4.
//
#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0),left(nullptr),right(nullptr) {}
    TreeNode(int x) : val(x),left(nullptr),right(nullptr) {}
    TreeNode(int x,TreeNode *left,TreeNode* right):val(x),left(left),right(right) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        //检测合法BST，可用中序遍历来检测
        TreeNode* pre = nullptr;
        return inOrder(root,pre);
    }
    vector<string> printTree(TreeNode* root){
        string s;
        vector<string> levelOrder = levelOrderStr(root);
        return levelOrder;
    }
    void deleteTree(TreeNode* root){
        if(root){
            deleteTree(root->left);
            deleteTree(root->right);
            root = nullptr;
        }
    }
    vector<string> levelOrderStr(TreeNode* root){
        vector<TreeNode*> cur;
        vector<TreeNode*> next;
        if(!root)
            return vector<string> {};
        if(root)
            cur.push_back(root);
        vector<string> res;
//        vector<string> curEles;
        while(cur.size()){
            //标识next是否都为null
            int flag = 0;
            for(int i = 0;i < cur.size();i++){
                TreeNode* node = cur[i];
                if(!node){
                    res.push_back("null");
                    continue;
                }
                else
                    res.push_back(to_string(node->val));
                //如果node的孩子不为null,则改变flag
                if(node->left || node->right)
                    flag++;
                next.push_back(node->left);
                next.push_back(node->right);
            }
            cur = next;
            if(flag == 0)
                cur.clear();
            next.clear();
        }
        return res;
    }
private:
    bool inOrder(TreeNode* root,TreeNode*& pre){
        if(root){
            bool leftValid = inOrder(root->left,pre);
            if(!leftValid)
                return false;
            if(!pre)
                pre = root;
                //注意这里如果存在相同节点,则返回false
                //BST不应该存在相同元素
            else if(pre->val >= root->val)
                return false;
            else{
                //pre->val < root->val
                pre = root;
            }
            bool rightValid = inOrder(root->right,pre);
            if(!rightValid)
                return false;
        }
        return true;
    }
};

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main() {
    TreeNode* root = new TreeNode(5);
    TreeNode* left = new TreeNode(1);
    TreeNode* right = new TreeNode(4);
    TreeNode* ll = nullptr;
    TreeNode* lr = nullptr;
    TreeNode* rl = new TreeNode(3);
    TreeNode* rr = new TreeNode(6);
    root->left = left;
    root->right = right;
    left->left = ll;
    left->right = lr;
    right->left = rl;
    right->right = rr;
    bool res = Solution().isValidBST(root);
    vector<string> rootArr = Solution().levelOrderStr(root);
    //[5,1,4,null,null,3,6]
    printVec(rootArr);
    cout<<boolalpha<<res<<endl;
    //result
    //false;
}
